text/state-passing.md

```diff
@@ -0,0 +1,156 @@
+- Title: Threading the state in the implementation
+- Driver: Hugo
+- Status: Draft
+
+# Summary and motivation
+
+Coq keeps internal data in tables updated by side effects. Several bugs or hacks come from the difficulty to control the flow of these side effects. This ceps proposes some directions to thread tables updating in a purely functional way.
+
+# Current structure of state in Coq
+
+All global tables are registered to `summary.ml` and `Summary.freeze_summaries : marshallable -> Summary.frozen`, to a few exceptions:
+- The list of entries of a document is in `Lib.lib_state` and not registered as a summary
+- The list of open goals is in `Proof_global.pstates : Proof_global.t` and is not synchronized
+
+The combination of all summaries and `lib_state` is called `States.state`. The combination of `States.state` and `Proof_global.t` is in `Vernacstate.t`. It can be thought as _the_ global state of Coq.
+
+# General proposal
+
+## State-passing-translation / environment-passing-translation
+
+The proposal is to differentiate between two kinds of functions.
+- Functions manipulating the state generically shall be "state-passing-translated" (or "environment-passing-translated" depending on need) over the type `States.state`
+- Functions morally manipulating only a reduced number of components of the global state shall be "state-passing-translated" (or "environment--passing-translated" depending on need) over an ad hoc subset of `States.state` (e.g. `printing_state`, or `typing_state`, or `tactic_state`, ...)
+
+## Registering of tables/summaries
+
+The proposal is to reverse the model of registration of a table. Instead of declaring a local mutable object to the summary manager by sending to it read/write/init methods (in practice called `freeze`/`unfreeze`/`init`), we ask the summary manager to send us read and modify methods on the global state (so-called "lenses" as far as I understood).
+
+# Key components managing states
+
+## Declaring a global table: `declare_summary`
+
+We shall rename `Summaries.frozen` into `Summaries.summaries` to hightlight that it is not anymore a "freeze", but _the_ (functional) summary of all tables.
+
+We shall provide instead a registration method (see #8624)
+```
+val declare_functional_summary : name:string -> 'a -> 'a Dyn.tag
```

Comment:
Shouldn't this modify the summaries? ie be `'a -> summaries -> summaries * 'a tag`
If not what is done with the `'a` argument?